### PR TITLE
Inversion accumulator

### PR DESCRIPTION
**Pull request by akoryagin**
- add two accumulators: `WeightedAccumulator` and `RegressionAccumulator`
- add `SyntheticGenerator`-tutorial in tutorials' folder; add section that explains how parameters influence resulting synthetic
- slightly refactor `SyntheticGenerator`-class

### CODE REVIEW

File: seismiqb/src/utils/classes.py [hunk 1]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+    def _update(self, crop, location):
+        # Scale incoming crop to better fit already filled data
+        # Fit is done via least-squares regression
+        overlap_data = self.data[location]
+        overlap_weights = self.weights[location]
+        crop_weights = self.kernel(crop.shape)
+
+        # If some of the values are already filled, use regression to fit new crop
+        # to what's filled
+        ixs_filled = np.where((~np.isnan(overlap_data)) & (~np.isnan(crop)))
```
[the remark this answers is not in the document]
Comment (SergeyTsimfer) [reply seismiqb/src/utils/classes.py:578]: `previous_data`, `previous_weights`, `overlap_indices`, `new_indices`?

File: seismiqb/src/utils/classes.py [hunk 2]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+        self.weights[location] += crop_weights
+
+    def _aggregate(self):
+        # Cleanup
+        self.remove_placeholder('weights')
+
+class RegressionAccumulator(Accumulator3D):
+    """ Accumulator that fits least-squares regression to scale values of
+    each incoming crop to match values of the overlap. In doing so, ignores nan-values.
+    For aggregation uses weighted sum of crops. Weights-making for crops is controlled by
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/classes.py:554]: Formatting

File: seismiqb/src/utils/classes.py [hunk 3]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+    `kernel`-parameter. """
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/classes.py:555]: Add the fact that this accumulator relies on the sequence of data arrival and can change the outcome depending on it

File: seismiqb/src/utils/classes.py [hunk 4]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+            # Update location-slice with weighed average
+            overlap_data[ixs_filled] = ((overlap_weights[ixs_filled] * overlap_data[ixs_filled]
+                                         + crop_weights[ixs_filled] * crop[ixs_filled]) /
+                                        (overlap_weights[ixs_filled] + crop_weights[ixs_filled]))
+
+            # Update weights over overlap
+            overlap_weights[ixs_filled] += crop_weights[ixs_filled]
+
+            # Use values from crop to update the region covered by the crop
+            # and not yet filled
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/classes.py:601]: Why two lines?

File: seismiqb/src/utils/classes.py [hunk 5]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+            overlap_data[ixs_new] = crop[ixs_new]
+            overlap_weights[ixs_new] = crop_weights[ixs_new]
+        else:
+            overlap_data[...] = crop
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/classes.py:605]: I would rather use the same lines, as in all previous accumulators: `self.data[location] = crop`, `self.data[location][overlap_indices] = ...`

File: seismiqb/src/utils/functions.py [hunk 6]
```diff
@@ -354,6 +354,27 @@ def make_bezier_figure(n=7, radius=0.2, sharpness=0.05, scale=1.0, shape=(1, 1),
...
+    to alpha on borders.
+    """
+    result = np.zeros(length)
+    array = np.linspace(alpha, 2, length)
+    result[:length // 2] = array[:length // 2]
+    result[length // 2:] = 2 + alpha - array[length // 2:]
+    return result
+
+def linear_kernel_nd(shape, alpha=.1):
+    """ Make kernel that decreases linearly from a center point to alpha on borders.
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/functions.py:368]: Formatting

File: seismiqb/src/utils/functions.py [hunk 7]
```diff
@@ -354,6 +354,27 @@ def make_bezier_figure(n=7, radius=0.2, sharpness=0.05, scale=1.0, shape=(1, 1),
     figure_coordinates = np.unique(np.ceil(curve_segments).astype(int), axis=0)
     return figure_coordinates
 
+def linear_kernel_1d(length, alpha):
+    """ Make linear kernel that decreases linearly from a center point
```
Comment (SergeyTsimfer) [inline seismiqb/src/utils/functions.py:358]: Formatting

Comment (SergeyTsimfer) [reply seismiqb/src/utils/functions.py:358]: It is not restricted to `alpha < 1`, so it is not always decreasing, right?

File: seismiqb/src/synthetic/generator.py [hunk 8]
```diff
@@ -101,6 +101,13 @@ def make_velocities(self, num_reflections=200, vel_limits=(900, 5400), horizon_h
         self.num_reflections = num_reflections
         return self
 
+    def make_upward_velocities(self):
+        """ Make array of upward velocities (with only positive diffs) out of existing array of velocities.
+        """
+        v0 = self.velocities[0]
+        self.upward_velocities = np.cumsum([v0] + np.abs(np.diff(self.velocities)).tolist())
```
Comment (SergeyTsimfer) [inline seismiqb/src/synthetic/generator.py:108]: `np.cumsum(np.abs(np.diff(self.velocities, prepend=2*v0)))`

File: seismiqb/src/utils/classes.py [hunk 9]
```diff
@@ -521,6 +522,93 @@ def _aggregate(self):
...
+
+        self.create_placeholder(name='data', dtype=self.dtype, fill_value=0)
+        self.create_placeholder(name='weights', dtype=np.float32, fill_value=0)
+        self.kernel = kernel
+
+    def _update(self, crop, location):
+        # Weights matrix for the incoming crop
+        crop_weights = self.kernel(crop.shape)
+        self.data[location] = ((crop_weights * crop + self.data[location] * self.weights[location]) /
+                                crop_weights + self.weights[location])
```
[the remark this answers is not in the document]
Comment (akoryagin) [reply seismiqb/src/utils/classes.py:544]: Nope, that's a bug

**Review by SergeyTsimfer — APPROVED** [review]

**Pull request merged**